lib/Interpreter/CIFactory.cpp

```diff
@@ -338,9 +337,442 @@ static bool getVisualStudioDir(std::string &path) {
   return false;
 }
 
-#endif // _MSC_VER
+#elif defined(__APPLE__)
+
+#include <dlfcn.h> // dlopen to avoid linking with CoreServices
+#include <CoreServices/CoreServices.h>
+#include <sstream>
+
+static bool getISysRootVersion(const std::string& SDKs, int major,
+                               int minor, std::string& sysRoot) {
+  std::ostringstream os;
+  os << SDKs << "MacOSX" << major << "." << minor << ".sdk";
+
+  std::string SDKv = os.str();
+  if (llvm::sys::fs::is_directory(SDKv)) {
+    sysRoot.swap(SDKv);
+    return true;
+  }
+
+  return false;
+}
+
+static bool getISysRoot(std::string& sysRoot) {
+  using namespace llvm::sys;
+
+  // Some versions of OS X and Server have headers installed
+  if (fs::is_regular_file("/usr/include/stdlib.h"))
+    return false;
+
+  std::string SDKs("/Applications/Xcode.app/Contents/Developer");
+
+  // Is XCode installed where it usually is?
+  if (!fs::is_directory(SDKs)) {
+    // Nope, use xcode-select -p to get the path
+    if (FILE *pf = ::popen("xcode-select -p", "r")) {
+      SDKs.clear();
+      char buffer[512];
+      while (fgets(buffer, sizeof(buffer), pf) && buffer[0])
+        SDKs.append(buffer);
+
+      // remove trailing \n
+      while (!SDKs.empty() && SDKs.back() == '\n')
+        SDKs.resize(SDKs.size() - 1);
+      ::pclose(pf);
+    } else // Nothing more we can do
+      return false;
+  }
+
+  SDKs.append("/Platforms/MacOSX.platform/Developer/SDKs/");
+  if (!fs::is_directory(SDKs))
+    return false;
+
+
+  // Seems to make more sense to get the currently running SDK so any loaded
+  // libraries won't casue conflicts
+
+  // Try to get the SDK for whatever version of OS X is currently running
+  if (void *core = dlopen(
+          "/System/Library/Frameworks/CoreServices.framework/CoreServices",
+          RTLD_LAZY)) {
+    SInt32 majorVersion = -1, minorVersion = -1;
+    typedef ::OSErr (*GestaltProc)(::OSType, ::SInt32 *);
+    if (GestaltProc Gestalt = (GestaltProc)dlsym(core, "Gestalt")) {
+      Gestalt(gestaltSystemVersionMajor, &majorVersion);
+      Gestalt(gestaltSystemVersionMinor, &minorVersion);
+    }
+    ::dlclose(core);
+
+    if (majorVersion != -1 && minorVersion != -1) {
+      if (getISysRootVersion(SDKs, majorVersion, minorVersion, sysRoot))
+        return true;
+    }
+  }
+
+#define GET_ISYSROOT_VER(maj, min) \
+  if (getISysRootVersion(SDKs, maj, min, sysRoot)) \
+    return true;
+
+  // Try to get the SDK for whatever cling was compiled with
+  #if defined(MAC_OS_X_VERSION_10_11)
+    GET_ISYSROOT_VER(10, 11);
+  #elif defined(MAC_OS_X_VERSION_10_10)
+    GET_ISYSROOT_VER(10, 10);
+  #elif defined(MAC_OS_X_VERSION_10_9)
```

Comment:
I agree we should have a configure time check. It covers only the cases when building cling on the same machine on which it will run.

What about if I take a prebuilt binary and run it on 10.0? How cling would warn me? There is no configure time there, right? Maybe this code should be moved in cling's RuntimeUnverse.h (which gets #included every time cling starts up).